homeassistant/components/firmata/__init__.py

```diff
@@ -0,0 +1,175 @@
+    {
+        vol.Required(CONF_NAME): cv.string,
+        vol.Required(CONF_PIN): vol.Any(cv.positive_int, ANALOG_PIN_SCHEMA),
+        # will be analog mode in future too
+        vol.Required(CONF_PIN_MODE): vol.Any(PIN_MODE_INPUT, PIN_MODE_PULLUP),
+        vol.Optional(CONF_NEGATE_STATE, default=False): cv.boolean,
+    },
+    required=True,
+    extra=vol.ALLOW_EXTRA,
+)
+
+BOARD_CONFIG_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_SERIAL_PORT): cv.string,
+        vol.Optional(CONF_SERIAL_BAUD_RATE): cv.positive_int,
+        vol.Optional(CONF_ARDUINO_INSTANCE_ID): cv.positive_int,
+        vol.Optional(CONF_ARDUINO_WAIT): cv.positive_int,
+        vol.Optional(CONF_SLEEP_TUNE): vol.All(
+            vol.Coerce(float), vol.Range(min=0.0001)
+        ),
+        vol.Optional(CONF_SAMPLING_INTERVAL): cv.positive_int,
+        vol.Optional(CONF_SWITCHES): [SWITCH_SCHEMA],
+        vol.Optional(CONF_BINARY_SENSORS): [BINARY_SENSOR_SCHEMA],
+    },
+    required=True,
+    extra=vol.ALLOW_EXTRA,
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.All(cv.ensure_list, [BOARD_CONFIG_SCHEMA])}, extra=vol.ALLOW_EXTRA
+)
+
+
+async def async_setup(hass, config):
+    """Set up the Firmata domain."""
+    # Delete all entries if domain removed from config
+    if DOMAIN not in config:
+        if hass.config_entries.async_entries(DOMAIN):
+            for entry in hass.config_entries.async_entries(DOMAIN):
+                await hass.config_entries.async_remove(entry.entry_id)
+        return True
+
+    # Delete specific entries that no longer exist in the config
+    if hass.config_entries.async_entries(DOMAIN):
+        for entry in hass.config_entries.async_entries(DOMAIN):
+            remove = True
+            for board in config[DOMAIN]:
+                if entry.data[CONF_SERIAL_PORT] == board[CONF_SERIAL_PORT]:
+                    remove = False
+                    break
+            if remove:
+                await hass.config_entries.async_remove(entry.entry_id)
+
+    # Setup new entries and update old entries
+    for board in config[DOMAIN]:
+        firmata_config = copy(board)
+        existing_entry = False
+        for entry in hass.config_entries.async_entries(DOMAIN):
+            if board[CONF_SERIAL_PORT] == entry.data[CONF_SERIAL_PORT]:
+                existing_entry = True
+                firmata_config[CONF_NAME] = entry.data[CONF_NAME]
+                hass.config_entries.async_update_entry(entry, data=firmata_config)
+                break
+        if not existing_entry:
+            hass.async_create_task(
+                hass.config_entries.flow.async_init(
+                    DOMAIN,
+                    context={"source": config_entries.SOURCE_IMPORT},
+                    data=firmata_config,
+                )
+            )
+
+    async def handle_shutdown(event):
+        """Handle shutdown of baords when Home Assistant shuts down."""
+        for board in hass.data[DOMAIN]:
```

Comment:
Just shutdown the corresponding board that is registered in `async_setup_entry`.